pkg/smartcontract/nef/nef.go

```diff
@@ -0,0 +1,171 @@
+type Version struct {
+	Mayor    int32
+	Minor    int32
+	Build    int32
+	Revision int32
+}
+
+// NewFile returns new NEF3 file with script specified.
+func NewFile(script []byte) File {
+	file := File{
+		Header: Header{
+			Magic:    magic,
+			Compiler: "neo-go",
+			Version: Version{
+				Mayor:    1, // TODO
+				Minor:    1,
+				Build:    1,
+				Revision: 1,
+			},
+			ScriptHash: hash.Hash160(script),
+		},
+		Script: script,
+	}
+	file.Checksum = file.Header.CalculateChecksum()
+	return file
+}
+
+// EncodeBinary implements io.Serializable interface.
+func (v *Version) EncodeBinary(w *io.BinWriter) {
+	w.WriteU32LE(uint32(v.Mayor))
+	w.WriteU32LE(uint32(v.Minor))
+	w.WriteU32LE(uint32(v.Build))
+	w.WriteU32LE(uint32(v.Revision))
+}
+
+// DecodeBinary implements io.Serializable interface.
+func (v *Version) DecodeBinary(r *io.BinReader) {
+	v.Mayor = int32(r.ReadU32LE())
+	v.Minor = int32(r.ReadU32LE())
+	v.Build = int32(r.ReadU32LE())
+	v.Revision = int32(r.ReadU32LE())
+}
+
+// EncodeBinary implements io.Serializable interface.
+func (h *Header) EncodeBinary(w *io.BinWriter) {
+	w.WriteU32LE(h.Magic)
+	w.WriteFixedString(h.Compiler, compilerFieldSize)
+	h.Version.EncodeBinary(w)
+	h.ScriptHash.EncodeBinary(w)
+}
+
+// DecodeBinary implements io.Serializable interface.
+func (h *Header) DecodeBinary(r *io.BinReader) {
+	h.Magic = r.ReadU32LE()
+	if h.Magic != magic {
+		r.Err = errors.New("invalid magic")
+		return
+	}
+	h.Compiler = r.ReadFixedString(compilerFieldSize)
+	h.Version.DecodeBinary(r)
+	h.ScriptHash.DecodeBinary(r)
+}
+
+// CalculateChecksum returns first 4 bytes of SHA256(Header) converted to uint32.
+func (h *Header) CalculateChecksum() uint32 {
+	buf := io.NewBufBinWriter()
+	h.EncodeBinary(buf.BinWriter)
+	if buf.Err != nil {
+		panic(buf.Err)
+	}
+	return binary.LittleEndian.Uint32(hash.Sha256(buf.Bytes()).BytesBE())
+}
+
+// EncodeBinary implements io.Serializable interface.
+func (n *File) EncodeBinary(w *io.BinWriter) {
+	n.Header.EncodeBinary(w)
+	w.WriteU32LE(n.Checksum)
+	w.WriteVarBytes(n.Script)
+}
+
+// DecodeBinary implements io.Serializable interface.
+func (n *File) DecodeBinary(r *io.BinReader) {
+	n.Header.DecodeBinary(r)
+	n.Checksum = r.ReadU32LE()
+	checksum := n.Header.CalculateChecksum()
+	if checksum != n.Checksum {
+		r.Err = errors.New("CRC verification fail")
+		return
+	}
+	n.Script = r.ReadVarBytes()
+	if len(n.Script) > maxScriptLength {
+		r.Err = errors.New("invalid script length")
+		return
+	}
+	if !hash.Hash160(n.Script).Equals(n.Header.ScriptHash) {
+		r.Err = errors.New("script hashes mismatch")
+		return
+	}
+}
+
+// Bytes returns byte array with serialized NEF File.
+func (n File) Bytes() ([]byte, error) {
+	buf := io.NewBufBinWriter()
+	n.EncodeBinary(buf.BinWriter)
+	if buf.Err != nil {
+		return nil, buf.Err
+	}
+	return buf.Bytes(), nil
+}
+
+// FromBytes returns NEF File deserialized from given bytes.
+func FromBytes(source []byte) (File, error) {
```

Comment:
Either it should be `FileFromBytes` or we need a `New` instead of `NewFile`.

Reply:
Replaced this with `FileFromBytes`.